src/topology/mod.rs

```diff
@@ -0,0 +1,428 @@
+pub mod server;
+pub mod monitor;
+
+use Error::{self, ArgumentError, OperationError};
+use Result;
+
+use bson::oid;
+
+use common::ReadPreference;
+use connstring::{ConnectionString, Host};
+use pool::PooledStream;
+
+use rand::{thread_rng, Rng};
+
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::{Arc, RwLock};
+use std::sync::atomic::AtomicIsize;
+use std::thread;
+
+use self::server::{Server, ServerDescription, ServerType};
+
+const DEFAULT_HEARTBEAT_FREQUENCY_MS: u32 = 10000;
+const MAX_SERVER_RETRY: usize = 3;
+
+/// Describes the type of topology for a server set.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum TopologyType {
+    Single,
+    ReplicaSetNoPrimary,
+    ReplicaSetWithPrimary,
+    Sharded,
+    Unknown,
+}
+
+/// Topology information gathered from server set monitoring.
+#[derive(Clone)]
+pub struct TopologyDescription {
+    pub ttype: TopologyType,
```

Comment:
I'd recommend changing this name to something more semantic like `topology_type`. Makes code easier to read.